[github action] Update github PRs with links to linear issues

escherize opened the pull request:
fixes https://github.com/metabase/metabase/issues/54821
closes dev-363

this PR is synced with a linear issue. The linear issue is linked to 2 issues: one in metabase/metabase, and one in a fake metabase repo and this bot auto-links the metabase/metabase issue only.

delete or modify the first line of this comment.

Now it'll get links automatically ✨

escherize committed:
first crack at linear pr link bot

escherize committed:
try it again

escherize committed:
bump actions versions

iethree commented on .github/workflows/linear-pr-links.yml:
```diff
@@ -0,0 +1,104 @@
+name: Update PR Description with Linear Link
+
+on:
+  push:
+    branches:
+      - '*'  # Triggers on push to any branch that has an open PR
```
I don't think this is the right trigger
- I think this will trigger on any push to any branch, whether or not it has a pr (like master or release branches)
- if you trigger on push, then open the PR, it won't ever be able to edit the description (since it doesn't exist yet)

escherize replied:
👍

iethree commented on .github/workflows/linear-pr-links.yml:
```diff
@@ -0,0 +1,104 @@
+  update-pr-description:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
you want checkout@v4, v3 is deprecated

iethree commented on .github/workflows/linear-pr-links.yml:
```diff
@@ -0,0 +1,104 @@
+      - uses: actions/checkout@v3
+
+      - name: Find Linear issues and update PR description
+        uses: actions/github-script@v6
```
you want @v7, @v6 will get deprecated soon if it hasn't already

escherize replied:
👍

iethree commented on .github/workflows/linear-pr-links.yml:
```diff
@@ -0,0 +1,104 @@
+          script: |
+            const branchName = context.payload.ref.replace('refs/heads/', '');
+
+            // Get PRs associated with this branch
```
I think you can do it the other way around, get the branchName from the PR event?

escherize committed:
Merge remote-tracking branch 'origin/master' into dev-363-auto-link-closed-github-issues

escherize committed:
bump

escherize committed:
try again

Added Node.js setup step
Added dependency installation step for node-fetch
Changed the ES module import to a CommonJS require statement
Changed variable name from fetch to nodeFetch for clarity

escherize committed:
extract script

escherize committed:
use the right graphql query

escherize committed:
add debug

escherize committed:
look at the right part of the linear api response

escherize committed:
lets see the issue numbers

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
again

escherize committed:
fix formatting

escherize committed:
remove ???'s

escherize committed:
improve logging

escherize commented on .github/scripts/update-pr-with-issues.js:
```diff
@@ -0,0 +1,173 @@
+
+    let shouldUpdate = false;
+    for (const num of issueNumbers) {
+      if (!newBody.includes(`closes #${num}`) && !newBody.includes(`Closes #${num}`)) {
```
Do you think we should we check all the magic keywords?

escherize replied:
done

escherize committed:
check all magic closing words

escherize committed:
refine checking logic

iethree committed:
test

iethree commented on .github/scripts/update-pr-with-issues.js:
```diff
@@ -0,0 +1,178 @@
+        body: newBody
+      });
+
+      await httpsRequest({
+        hostname: 'api.github.com',
+        path: `/repos/${repoOwner}/${repoName}/pulls/${prNumber}`,
+        method: 'PATCH',
+        headers: {
+          'Accept': 'application/vnd.github.v3+json',
+          'Authorization': `token ${githubToken}`,
+          'User-Agent': 'Linear-GitHub-Action',
+          'Content-Type': 'application/json',
+          'Content-Length': updateData.length
+        }
+      }, updateData);
```
I also managed to break this

https://github.com/metabase/metabase/actions/runs/13730843983/job/38407486203?pr=54779

iethree commented on .github/scripts/update-pr-with-issues.js:
```diff
@@ -0,0 +1,178 @@
+const https = require('https');
+
+function containsIssueReference(text, issueNumber) {
+  const regex = new RegExp(`(close|closes|closed|fix|fixes|fixed|resolve|resolves|resolved):?\\s*#?${issueNumber}\\b`, 'i');
+  return regex.test(text);
+}
```
- this will catch issues referenced like this`closes #123` 
- but not `closes https://github.com/metabase/metabase/issues/123`

there's a better regex in `linked-issues.ts` that you can probably adapt

iethree reviewed (changes requested):
Looks good! very nice find with `attachmentsForURL` 

found a couple bugs though

escherize committed:
some review responses

escherize committed:
again

escherize committed:
make github object a thing

escherize committed:
install octokit

escherize committed:
install octokit again

escherize committed:
install octokit again

escherize committed:
running

escherize committed:
running

escherize committed:
running

escherize committed:
running

escherize committed:
running

escherize committed:
delimit message with spaces

escherize committed:
delimit message with spaces

escherize committed:
delimit message with spaces

escherize committed:
delimit message with spaces

escherize committed:
test

escherize committed:
only try to close metabase/metabase issues

iethree reviewed (approved):
Nice! tried to break it, couldn't 👍
